pkg/didcomm/common/service/did_comm_msg.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package service
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/decorator"
+)
+
+const (
+	jsonID       = "@id"
+	jsonType     = "@type"
+	jsonThread   = "~thread"
+	jsonThreadID = "thid"
+)
+
+// Header helper structure which keeps reusable fields
+// Deprecated: Use DIDCommMsg instead of Header
+// TODO: Remove deprecated Header structure https://github.com/hyperledger/aries-framework-go/issues/1075
+type Header struct {
+	ID     string           `json:"@id"`
+	Thread decorator.Thread `json:"~thread"`
+	Type   string           `json:"@type"`
+	// TODO revisit ~purpose, should be generic map [Issue #1037]
+	Purpose []string `json:"~purpose"`
+}
+
+// DIDCommMsg did comm msg
+type DIDCommMsg map[string]interface{}
+
+// NewDIDCommMsg returns DIDCommMsg with Header
+func NewDIDCommMsg(payload []byte) (DIDCommMsg, error) {
+	var msg DIDCommMsg
+
+	err := json.Unmarshal(payload, &msg)
+	if err != nil {
+		return nil, fmt.Errorf("invalid payload data format: %w", err)
+	}
+
+	return msg, nil
+}
+
+// ThreadID returns msg ~thread.thid if there is no ~thread.thid returns msg @id
+// message is invalid if ~thread.thid exist and @id is absent
+func (m DIDCommMsg) ThreadID() (string, error) {
+	if m == nil {
+		return "", ErrInvalidMessage
+	}
+
+	msgID := m.ID()
+	thread, ok := m[jsonThread].(map[string]interface{})
+
+	if ok && thread[jsonThreadID] != nil {
+		var thID string
+		if v, ok := thread[jsonThreadID].(string); ok {
+			thID = v
+		}
+
+		// if message has ~thread.thid but @id is absent this is invalid message
+		if len(thID) > 0 && msgID == "" {
+			return "", ErrInvalidMessage
+		}
+
+		if len(thID) > 0 {
+			return thID, nil
+		}
+	}
+
+	// we need to return it only if there is no ~thread.thid
+	if len(msgID) > 0 {
+		return msgID, nil
+	}
+
+	return "", ErrThreadIDNotFound
+}
+
+// Type returns the message type
+func (m DIDCommMsg) Type() string {
+	if m == nil || m[jsonType] == nil {
+		return ""
+	}
+
+	res, ok := m[jsonType].(string)
+	if !ok {
+		return ""
+	}
+
+	return res
+}
+
+// ID returns the message id
+func (m DIDCommMsg) ID() string {
+	if m == nil || m[jsonID] == nil {
+		return ""
+	}
+
+	res, ok := m[jsonID].(string)
+	if !ok {
+		return ""
+	}
+
+	return res
+}
+
+// Scan converts message to  struct
+func (m DIDCommMsg) Scan(v interface{}) error {
```

Comment:
What does `Scan` mean for the user of this API? 

Personally, I prefer `DecodeTo` as portrayed by the function logic or `Convert` as per the doc comment.

or perhaps could it be renamed to `Unmarshal(v)` or `CopyTo(v)` ?  

The function name should be easy to guess and accurate [ref](https://talks.golang.org/2014/names.slide#3), `Scan` can mean different things to different people.

Reply:
From the code usage refactored in other files below, this function is used exactly like json.Unmarshal() ... so probably `Unmarshal()` makes more sense.

Reply:
I also had a bit trouble reconciling the word `Scan` in this usage.

Also note the underlying `mapstructure` library seems to call it Decode (https://godoc.org/github.com/mitchellh/mapstructure#Decode)